app/src/lib/git/for-each-ref.ts

```diff
@@ -101,3 +101,104 @@ export async function getBranches(
 
   return branches
 }
+
+/**
+ * Gets all branches that differ from their upstream (i.e. they're ahead,
+ * behind or both), excluding the current branch.
+ * Useful to narrow down a list of branches that could potentially be fast
+ * forwarded.
+ *
+ * @param repository Repository to get the branches from.
+ * @param allBranches All known branches in the repository.
+ * @param currentBranchName Name of the current branch.
+ */
+export async function getBranchesDifferingFromUpstream(
+  repository: Repository,
+  allBranches: ReadonlyArray<Branch>,
+  currentBranchName: string | null
+): Promise<ReadonlyArray<Branch>> {
+  const format = [
+    '%(refname)',
+    '%(refname:short)',
+    '%(objectname)', // SHA
+    '%(upstream)',
+    '%(symref)',
+  ].join('%00')
+
+  const prefixes = ['refs/heads', 'refs/remotes']
+
+  const result = await git(
+    ['for-each-ref', `--format=${format}`, ...prefixes],
+    repository.path,
+    'getBranchesDifferingFromUpstream',
+    { expectedErrors: new Set([GitError.NotAGitRepository]) }
+  )
+
+  if (result.gitError === GitError.NotAGitRepository) {
+    return []
+  }
+
+  const lines = result.stdout.split('\n')
+
+  // Remove the trailing newline
+  lines.splice(-1, 1)
+
+  if (lines.length === 0) {
+    return []
+  }
+
+  const localBranches = []
+  const remoteBranchShas = new Map<String, String>()
+
+  // First we need to collect the relevant info from the command output:
+  // - For local branches with upstream: name, ref, SHA and the upstream.
+  // - For remote branches we only need the sha (and the ref as key).
+  for (const line of lines) {
+    const pieces = line.split('\0')
+
+    const ref = pieces[0]
+    const name = pieces[1]
+    const sha = pieces[2]
+    const upstream = pieces[3]
+    const symref = pieces[4]
```

Comment:
We can clean this up a tad by deconstructing

```suggestion
    const [ref, name, sha, upstream, symref] = line.split('\0')
```